alicloud/resource_alicloud_ess_notification.go

```diff
@@ -0,0 +1,148 @@
+}
+
+func resourceAlicloudEssNotificationRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	essService := EssService{client}
+	object, err := essService.DescribeEssNotification(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+	d.Set("scaling_group_id", object.ScalingGroupId)
+	d.Set("notification_arn", object.NotificationArn)
+	d.Set("notification_types", object.NotificationTypes.NotificationType)
+	return nil
+}
+
+func resourceAlicloudEssNotificationUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	request := ess.CreateModifyNotificationConfigurationRequest()
+	scalingGroupId, notificationArn, err := parseEssNofiticationId(d.Id())
+	if err != nil {
+		return WrapError(err)
+	}
+	request.ScalingGroupId = scalingGroupId
+	request.NotificationArn = notificationArn
+	if d.HasChange("notification_types") {
+		v := d.Get("notification_types")
+		notificationTypes := notificationTypeSet2List(v.(*schema.Set))
+		if len(notificationTypes) > 0 {
+			request.NotificationType = &notificationTypes
+		}
+	}
+	raw, err := client.WithEssClient(func(essClient *ess.Client) (interface{}, error) {
+		return essClient.ModifyNotificationConfiguration(request)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
+	return resourceAlicloudEssNotificationRead(d, meta)
+}
+
+func resourceAlicloudEssNotificationDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	essService := EssService{client}
+	request := ess.CreateDeleteNotificationConfigurationRequest()
+	scalingGroupId, notificationArn, err := parseEssNofiticationId(d.Id())
+	if err != nil {
+		return WrapError(err)
+	}
+	request.ScalingGroupId = scalingGroupId
+	request.NotificationArn = notificationArn
+
+	raw, err := client.WithEssClient(func(essClient *ess.Client) (interface{}, error) {
+		return essClient.DeleteNotificationConfiguration(request)
+	})
+	if err != nil {
+		if IsExceptedErrors(err, []string{InvalidNotificationNotFound, InvalidScalingGroupIdNotFound}) {
+			return nil
+		}
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
+	return WrapError(essService.WaitForEssNotification(d.Id(), Deleted, DefaultTimeout))
+}
+
+func notificationTypeSet2List(notificationTypeSet *schema.Set) []string {
+	notificationTypes := make([]string, 0)
+	notificationTypeList := notificationTypeSet.List()
+	if len(notificationTypeList) > 0 {
+		for _, v := range notificationTypeList {
+			notificationTypes = append(notificationTypes, v.(string))
+		}
+	}
+	return notificationTypes
+}
+
+func parseEssNofiticationId(id string) (scalingGroupId string, notificationArn string, err error) {
```

Comment:
I think the method `parseEssNofiticationId` is redundant and just use `strings.SplitN` is ok.

Reply:
removed